### ISSUE TITLE
tensorflow-gpu using only 10% of my GPU

### ISSUE THREAD

**Issue by ismaelestalayo** [tensorflow/tensorflow#26758]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 (1809)
- TensorFlow installed from (source or binary): binary (pip install tensorflow-gpu)
- TensorFlow version (use command below): b'v1.10.0-rc1-19-g656e7a2b34' 1.10.0
- Python version: 3.6.8
- CUDA/cuDNN version: 9.0
- GPU model and memory: GTX 960 2GB

**Describe the current behavior**
While trying to train a neural network with my GTX960 after installing tensorflow-gpu, and choosing my GPU with the below code, I can see on the Windows task manager that it's only using about 10% of the GPU, and thus making it way slower than training it with the CPU. 

```
config = tf.ConfigProto()
config.gpu_options.allow_growth = False
config.gpu_options.per_process_gpu_memory_fraction = 1
session = tf.Session(config = config)

with tf.device("/device:GPU:0"):
    model = Sequential()
    ...
```

**Describe the expected behavior**
The GPU should be used almost if not entirely.

**Code to reproduce the issue**
you'd need the whole neural network code plus the datasets and the rest...

**Other info / logs**
>python predictor_LSTM_all.py
Using TensorFlow backend.
2019-03-15 17:38:18.822616: I T:\src\github\tensorflow\tensorflow\core\platform\cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2019-03-15 17:38:19.109114: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1405] Found device 0 with properties:
name: GeForce GTX 960 major: 5 minor: 2 memoryClockRate(GHz): 1.2785
pciBusID: 0000:01:00.0
totalMemory: 2.00GiB freeMemory: 1.64GiB
2019-03-15 17:38:19.117459: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1484] Adding visible gpu devices: 0
2019-03-15 17:38:19.527190: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-03-15 17:38:19.532341: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:971]      0
2019-03-15 17:38:19.535210: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:984] 0:   N
2019-03-15 17:38:19.538575: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 2048 MB memory) -> physical GPU (device: 0, name: GeForce GTX 960, pci bus id: 0000:01:00.0, compute capability: 5.2)
Epoch 1/10
2019-03-15 17:38:21.045228: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1484] Adding visible gpu devices: 0
2019-03-15 17:38:21.049900: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-03-15 17:38:21.054218: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:971]      0
2019-03-15 17:38:21.057053: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:984] 0:   N
2019-03-15 17:38:21.060264: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 2048 MB memory) -> physical GPU (device: 0, name: GeForce GTX 960, pci bus id: 0000:01:00.0, compute capability: 5.2)
15709/15709 [==============================] - 39s 2ms/step - loss: 33.3039 - acc: 0.1775

(using the CPU only takes about 7-10 seconds to finish the first epoch while here it shows how it takes almost 40 seconds with the GPU.

**Comment by tlkh** [comment]
What is the batch size you are using while training? Too small batch size makes GPU training inefficient.

In addition, are you using `tf.keras`? Are you using `LSTM` or `CuDNNLSTM`? The latter is much, much faster!

**Comment by ismaelestalayo** [comment]
> What is the batch size you are using while training? Too small batch size makes GPU training inefficient.

I was using a batch size of 20 as my dataset is 15700 samples long only. I just tried a batch size of 1000 and the GPU usage has gone up to 20%, so thanks for that, I had no idea. (I'm gonna have to rise the epochs as well if i want similar results tho)

> In addition, are you using `tf.keras`? Are you using `LSTM` or `CuDNNLSTM`? The latter is much, much faster!

Yes, I'm using Keras, and I'm currently using normal LSTMs so I'm going to do some research to thsoe CuDNNLSTM and see if I can modify my code slightly to use them, thanks again!

**Comment by ismaelestalayo** [comment]
wow @tlkh I'm in your debt for telling me about cuDNNLSTM. Not only it's faster than LSTM on CPU (which is going to save me quite some time over 111 sensors that I have to train and predict), but it's also going to save my poor CPU of quite a few rough hours.

**Comment by realiti4** [comment]
Can you check with GPU-z. I discovered last week that windows task manager shows max 14%, but it is 100% utilization most of the time on GPU-z.

**Comment by ismaelestalayo** [comment]
Yup @realiti4 , using GPU-z does show way higher GPU usage, and while it's only about 40% it probably is because i'm using a small Batch Size for the training.

![Sin título](https://user-images.githubusercontent.com/17100634/54489929-b4214000-48b1-11e9-80c5-b6ada1bfd021.png)

**Comment by ZIKO94ZIKO** [comment]
Hello @ismaelestalayo  , i have the some problem  ,when i run my script python ,  tensorflow-gpu using only 10% of my GPU , please tel me how can be resolve it , i see in you are using LSTM or CuDNNLSTM , please tell me how , thanks in advance.

**Comment by ZIKO94ZIKO** [comment]
Im using OS  Linux Ubuntu 18.04,  i have this problem since 2 week , please help me,

**Comment by ismaelestalayo** [comment]
> Hello @ismaelestalayo , i have the same problem ,when i run my script python , tensorflow-gpu using only 10% of my GPU , please tel me how can be resolve it , i see that you are using LSTM or CuDNNLSTM , please tell me how , thanks in advance.

Sorry to tell you that I did not "fix" it per se (if it even was a bug or error of Tensorflow). 
As you can see in my previous message, GPU-Z actually showed a 40% load on the GPU which is way more significant, and the other 60% missing could partly be because of very small batch sizes which are not optimal as another user suggested in this very issue. 

Lastly, if you're using LSTM, I recommend you trying CuDNNLSTM, as it's way more efficient and fast (while getting the same results as with the basic LSTMs).

**Comment by Shaul-Z** [comment]
Hi 
Thanks for bringing up this issue. When I try to increase the batch size over 14 the training process is crushing. Any idea why?

**Comment by ismaelestalayo** [comment]
> Hi
> Thanks for bringing up this issue. When I try to increase the batch size over 14 the training process is crushing. Any idea why?

First of all, what do you mean by "crushing"? And second, there's no fixed value that works for every dataset, the batch size is just the number of examples in each one of the epoch. The higher the batch size, the more memory you'll need

**Comment by Shaul-Z** [comment]
By crushing I mean training process is failing at the very start of it. 
GPU RTX 2080
Ram 32gib
I saw people with worse GPUs setting batch size to 1,000. I can set it to 14 and training process took 4 days to get to 1.5 in the loss function.

**Comment by ismaelestalayo** [comment]
> By crushing I mean training process is failing at the very start of it.
> GPU RTX 2080
> Ram 32gib
> I saw people with worse GPUs setting batch size to 1,000. I can set it to 14 and training process took 4 days to get to 1.5 in the loss function.

yep, it's weird to crush with such high specs, I'd suggest you Google the specific error that TensorFlow or Python returns. it might be specific to the size of your dataset not accepting high batch sizes or something, I can't tell just by the sheer specs alone.

**Comment by Ismail-Mangrio** [comment]
> By crushing I mean training process is failing at the very start of it.
> GPU RTX 2080
> Ram 32gib
> I saw people with worse GPUs setting batch size to 1,000. I can set it to 14 and training process took 4 days to get to 1.5 in the loss function.

I think its a memory overcommitment issue. The solution is to increase the memory size from the default settings. The following link might help you (it worked for me) :
https://stackoverflow.com/questions/57507832/unable-to-allocate-array-with-shape-and-data-type

**Comment by jbrepogmailcom** [comment]
I have similar problem, but with performance glitch. On low-end PC the step duration is 4ms. With added GPU NVIDIA T4, which is quite expensive, the GPU is used at 25% only and speed improvement is from 4ms per step to 3ms per step. It is improvement of only 25% when adding GPU that costs thousands of dollars?

**Comment by prettypositron** [comment]
> What is the batch size you are using while training? Too small batch size makes GPU training inefficient.
> 
> In addition, are you using `tf.keras`? Are you using `LSTM` or `CuDNNLSTM`? The latter is much, much faster!

I've been working on this for a while but raising my batch size fixed everything. Thank you very much!
